docs/design/proposals/codegen-refactor/model_cmd_2.md

```diff
@@ -0,0 +1,143 @@
+* `ackmodel`: the [code](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/model/model.go#L36) representation of ACK's view of the world; the source of truth for `aws-sdk`, `ackgenconfig`, and `reconciliation`
+
+## Problem
+There are issues that are starting to degrade the code generation `pipeline` experience and will compound as *code-generator* continues to grow:
+* Duplicate work between `api` and `controller` commands; both generate `ackmodel`
+* Overloaded functions such as [SetResource](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/generate/code/set_resource.go#L74); responsible for a lot more than generating Go code
+* Confusing CX. There's a not-so-clear dependency on a folder in the $SERVICE-controller via [getLatestAPIVersion](https://github.com/aws-controllers-k8s/code-generator/blob/26e5da2e7656bb836ee438c05df14f2adc50197d/cmd/ack-generate/command/common.go#L271). This goes against the flow of the `pipeline` and creates confusion as there is a `--version` flag for the `apis` command (the above function takes precedence).
+* Long dev loop; after making changes to either *generator.yaml* or logic in *code-generator* users need to either manually verify files or run e2e tests-- both of which are time consuming today
+* How *code-generator* reconciles `ackgenconfig` with `aws-sdk` to create `ackmodel` is inconsistent (and some duplicated work). For example, overrides such as `custom_shapes` [edit aws-sdk directly](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/sdk/custom_shapes.go#L62) while others such as `ignored_operations` [set operation values to nil](https://github.com/aws-controllers-k8s/code-generator/blob/26e5da2e7656bb836ee438c05df14f2adc50197d/pkg/model/model.go#L295) in memory leaving the actual operations in `aws-sdk` untouched.
+
+Current Pipeline
+---
+
+![current-pipeline](./images/current_pipeline.png)
+* Displayed are the important and mostly duplicated calls between the `apis` and `controller` commands
+* The cache is only used to store the `aws-sdk` repo
+* The red line denotes a dependency on $SERVICE-controller (output directory) repo that should be removed
+* Only `GetCRDs` is duplicated between the commands when creating the model
+
+
+Current Dev Loop
+---
+
+![current-dev-loop](./images/current_dev_loop.png)
+* Validating the result of a change in *code-generator* and/or *generator.yaml* can be a lengthy process
+* In the current feedback loop, users and contributors need to manually verify files or run lengthy e2e tests even for minor changes to `ackgenconfig`
+
+
+## Solution
+Add a new command, `model`, which takes in `aws-sdk` and *generator.yaml* as inputs to create and cache `ackmodel`. Existing commands, `api` and `controller`, will remove parameters for `aws-sdk` and *generator.yaml* and replace them with `ackmodel` input and/or let it default to `--cache-dir`.
```

Comment:
In your problem statement and "current dev loop", you mention the long validation process when changing the `generator.yaml` file. However, modifying the `generator.yaml` will bust the cache for this proposed `ackmodel` - meaning you would have to regenerate this cache every time anyway? This solution would only really be useful for quickly iterating on templates?

Furthermore, how much time does generating the `ackmodel` take? Assuming you are using the same AWS SDK version, it is probably less than a second to generate that model? Is this added complexity justified instead of caching `generator.yaml` files instead?

Reply:
> However, modifying the generator.yaml will bust the cache for this proposed ackmodel - meaning you would have to regenerate this cache every time anyway?

Good callout, I'll clarify in the next revision, but the intent is that the time savings is coming from being able to quickly compare `ackmodel`s between changes so users wouldn't need to follow longer validation processes involving building and running $SERVICE-controller. I had something like this in mind:
0. Set `cache-dir` to directory in code-generator, ex: `local-ack-cache`
1. Generate `ackmodel` using latest on *main*
2. Commit `local-ack-cache/ackmodel.json`
3. Change `generator.yaml`
4. Generate `ackmodel` with changed `generator.yaml`
5. `git diff local-ack-cache/ackmodel.json` to see changes


> This solution would only really be useful for quickly iterating on templates?

I agree that's mostly true for today. However, as the interface in `generator.yaml` becomes simpler and more powerful, 1-2 lines of config in `generator.yaml` may result in multiple complex relationships and overrides(stored in `ackmodel`) that may not be obvious. For platform devs, being able to quickly compare `generator.yaml` changes to relationships/overrides could mean quicker iteration for development and debugging.

> Furthermore, how much time does generating the ackmodel take? Assuming you are using the same AWS SDK version, it is probably less than a second to generate that model? Is this added complexity justified instead of caching generator.yaml files instead?

I don't have any performance metrics to compare, but also the small time savings by caching `ackmodel` isn't the focus. The biggest benefit to caching `ackmodel` is **consistency**:
* *Consistently* capture and provide the ACK state of the world to downstream commands like `generate apis` and `generate controller` and have confidence they generated controller artifacts based on the same source of truth. 


Caching *generator.yaml* only still leaves opportunity for compatibility issues:
  *  unintentionally changing `version` and `aws-sdk-go-version` flags between code-generator runs could impact artifacts despite *generator.yaml* being the same
  * generating apis, making an update to CRD creation logic, then generating controller would most likely result in incompatible types/syntax

These might be extreme examples, but hopefully shows the intent on reducing points of failure and possible inconsistencies (mainly for platform devs).

Reply:
Correct me if I'm wrong, but this is assuming an immutable `aws-sdk` being compared to an altered [Model.SDKAPI](https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/model/model.go#L37) which is not the case in this proposal. 

I'm proposing we modify the `aws-sdk-go` directly to adhere to `generator.yaml`'s "reality" so that `CreateBucketInputShape` in `Model.SDKAPI` would also be called `Name` instead of `Bucket` (using your example). If for whatever reason we cannot edit `aws-sdk-go` directly to match `generator.yaml`, then we can leverage new or existing fields in `ackmodel` such as `shapeRenames` (new!) or `typeRenames` to capture this override and resolve `API Inference` in places like `SetResource`:

If we cannot apply a config from `generator.yaml` to `aws-sdk-go` for reasons, we can use aux fields in `ackModel` to resolve the inference:

```
shape := m.GetOutputShape(outputShape)
for memberIndex, memberName := range shape.MemberNames() {
  // shape is the unaltered shape from aws-sdk-go
 // consult the model if API inference cannot be resolved
  if !memberFound {
    _, found := m.ShapeRenames[memberName]
   ...
}
}
```

The idea stems from if we need to do the work of applying `generator.yaml` to `aws-sdk-go` during model creation, then we should retain that knowledge somehow so we don't need to figure it out all over again in SetResource/SetSdk